### DISCUSSION TITLE
Heidke Skill Score Calculation Discrepancy

### DISCUSSION THREAD

**Discussion by lindsayrblank** [dtcenter/METplus#1964] [Statistical Computation] [answered]
Hello DTC folks,

I am using METviewer 4.1 to aggregate CTC and calculate the Heidke Skill Score. I saw some values that seemed unreasonable so I decided to calculate the score by hand. I used both the equation in the [UG](https://met.readthedocs.io/en/main_v10.0/Users_Guide/appendixC.html?highlight=heidke#heidke-skill-score-hss)  and that one I found in the [source code](https://github.com/dtcenter/MET/blob/3e43b63e0830e7b92d1bd59553c2832d0dfa1f85/met/src/libcode/vx_statistics/contable_stats.cc). I get a different answer than the one MET provides.

Here is the CTC I get from MET:
CTC_FN_ON	4370.0
CTC_FN_OY	40.0	
CTC_FY_ON	101.0
CTC_FY_OY	6.0

When I plug these values into both of the equations I mentioned above, I get an HSS of `0.0651149`. MET gives me an HSS of `0.9687771`. These are drastically different values and I can't figure out how to get the HSS MET produces. Can you please take a look and let me know if I am missing something? 

Thank you!
Lindsay

**Comment by JohnHalleyGotway** [comment]
@lindsayrblank I see you got some unexpected results from the computation of HSS reported by METviewer.

Thanks for sending sample data values to demonstrate. As a sanity check, I dropped them in to a sample
[ctc.txt](https://github.com/dtcenter/METplus/files/10044347/ctc.txt) file and ran it through stat_analysis to compute the corresponding CTS statistics:
```
stat_analysis -lookin ctc.txt -job aggregate_stat -line_type CTC -out_line_type CTS -out_stat cts.txt
```
And sure enough, the output [cts.txt](https://github.com/dtcenter/METplus/files/10044351/cts.txt) reports:
`HSS = 0.065115`.

So I wanted to check the corresponding equation in METviewer. The only trouble is that we're in the middle of a transition from the computation of statistics using R in METviewer over to the computation of statistics using python in METcalcpy.

1. First up, the [calcHSS](https://github.com/dtcenter/METviewer/blob/cbba3e0e0e2b13116244d546ba97659fd3ad9e41/R_work/include/statistics.R#L436) R function in METviewer version 4.1.
This does return the correct result:
```
R
calcHSS = function(d){
  if(is.na(d$fy_oy) || is.na(d$fn_oy) || is.na(d$total) ){
    return (NA);
  }
  if( 0 == d$total ){ return (NA); }
  dblC = ( ((d$fy_oy + d$fy_on) / d$total) *(d$fy_oy + d$fn_oy)  +  ((d$fn_oy + d$fn_on) / d$total) * (d$fy_on + d$fn_on)  ) ;
  hss = ( (d$fy_oy + d$fn_on - dblC) / (d$total - dblC) );
  return( round(hss, digits=5) );
}
d = data.frame("fy_oy"=6,"fy_on"=101,"fn_oy"=40,"fn_on"=4370,"total"=4517)
calcHSS(d)
[1] 0.06511
```
2. Next up, the [calculate_hss()](https://github.com/dtcenter/METcalcpy/blob/983c72b2bac9f53705b50d9f6373956cff6e85a3/metcalcpy/util/ctc_statistics.py#L371) python function in METcalcpy.
```
python
fy_oy,fy_on,fn_oy,fn_on,total = 6,101,40,4370,4517
dbl_c = ((fy_oy + fy_on) / total) * (fy_oy + fn_oy)
hss = ((fy_oy + fn_on - dbl_c) / (total - dbl_c))
hss
```
And that reports hss = 0.968777

So that's our culprit!

@TatianaBurek there is a bug in this [calculate_hss()](https://github.com/dtcenter/METcalcpy/blob/983c72b2bac9f53705b50d9f6373956cff6e85a3/metcalcpy/util/ctc_statistics.py#L371) in METcalcpy. Looks like the logic in R was not translated into python correctly. In particular, the "dbl_c" value is missing some terms.

Can you please write up a bugfix issue and provide a fix?

**Reply by TatianaBurek** [reply]
The issue is created : https://github.com/dtcenter/METcalcpy/issues/239

**Comment by TatianaBurek** [comment, marked as answer]
The new version of METcalcpy v1.1.2 with the fix for HSS is released:
https://github.com/dtcenter/METcalcpy/releases/tag/v1.1.2

https://github.com/dtcenter/METcalcpy/blob/4defd00d521c823c3e4b2a8fef4a0e0a35f325c5/metcalcpy/util/ctc_statistics.py#L371

**Comment by jprestop** [comment]
Hi @lindsayrblank.  I just wanted to follow up and see if the fix resolved the problem.  I’d like to mark this discussion as being answered and “lock it” to prevent future posts. That’s how we encourage users to create new discussions for new questions. But I wanted to give you an opportunity to comment on it before doing so. Please feel free to select one of the responses as being the best answer to your original question. That’ll help future users with similar questions find answers.

**Reply by lindsayrblank** [reply]
Hi @jprestop! This is my first day back in the office after a vacation, hence my delay. I have marked an answer and will test the solution at a later date (after I've caught up!). Thank you!

**Reply by jprestop** [reply]
Thanks @lindsayrblank.  I hope you had a great break and Thanksgiving!  I'll go ahead and lock this Discussion.  If you do experience any problems, please start a new Discussion and refer to this one if you'd like.
